core/src/test/java/com/google/cloud/sql/core/CloudSqlInstanceTest.java

```diff
@@ -119,18 +116,55 @@ public void testInstanceFailsOnConnectionError() throws Exception {
   }
 
   @Test
-  public void testCloudSqlInstanceForcesRefresh() throws InterruptedException {
+  public void testInstanceFailsOnTooLongToRetrieve() throws Exception {
+
+    InstanceDataSupplier instanceDataSupplier =
+        (CloudSqlInstanceName instanceName,
+            AccessTokenSupplier accessTokenSupplier,
+            AuthType authType,
+            ListeningScheduledExecutorService exec,
+            ListenableFuture<KeyPair> keyPair) -> {
+          return exec.submit(
+              () -> {
+                try {
+                  // stop for 2 minutes, which is longer than the timeout
+                  // to wait for getInstanceData()
+                  Thread.sleep(120 * 1000);
+                } catch (InterruptedException e) {
+                }
+                throw new RuntimeException("fake read timeout");
+              });
+        };
+
+    // initialize instance after mocks are set up
+    CloudSqlInstance instance =
+        new CloudSqlInstance(
+            "project:region:instance",
+            instanceDataSupplier,
+            AuthType.PASSWORD,
+            stubCredentialFactory,
+            executorService,
+            keyPairFuture,
+            RateLimiter.create(10));
+
+    RuntimeException ex = Assert.assertThrows(RuntimeException.class, instance::getSslData);
+    assertThat(ex).hasMessageThat().contains("No refresh has completed");
+  }
+
+  @Test
+  public void testCloudSqlInstanceForcesRefresh() throws Exception {
+    SslData sslData = new SslData(null, null, null);
+    InstanceData data =
+        new InstanceData(null, sslData, Date.from(Instant.now().plus(1, ChronoUnit.HOURS)));
     AtomicInteger refreshCount = new AtomicInteger();
 
     CloudSqlInstance instance =
         new CloudSqlInstance(
             "project:region:instance",
             (instanceName, accessTokenSupplier, authType, executor, keyPair) -> {
+              Thread.sleep(100);
```

Comment:
Can we avoid the sleep somehow? These are notoriously flaky otherwise.

Reply:
I've updated these tests (and added a few new ones) to use a lock instead of waiting.